### PR TITLE
gh-126476: Raise `IllegalMonthError` for calendar.formatmonth method when the input month is not corret

### LINKED ISSUE #126476: No proper validation for month in Calendar Module
# Bug report

### Bug description:

In the calendar module, `IllegalMonthError` is not handled properly. For months greater than 12, an IndexError is raised instead.

`formatmonthname ` do not raise IndexError for negative integer because list support negative integer indexing.

```python
>>> import calendar

>>> calendar.month(2024, -1)
Traceback (most recent call last):
  File "<python-input-12>", line 1, in <module>
    calendar.month(2024, -1)
    ~~~~~~~~~~~~~~^^^^^^^^^^
  File "/Users/gugu/Code/cpython/Lib/calendar.py", line 395, in formatmonth
    for week in self.monthdays2calendar(theyear, themonth):
                ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
  File "/Users/gugu/Code/cpython/Lib/calendar.py", line 284, in monthdays2calendar
    days = list(self.itermonthdays2(year, month))
  File "/Users/gugu/Code/cpython/Lib/calendar.py", line 240, in itermonthdays2
    for i, d in enumerate(self.itermonthdays(year, month), self.firstweekday):
                ~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/gugu/Code/cpython/Lib/calendar.py", line 228, in itermonthdays
    day1, ndays = monthrange(year, month)
                  ~~~~~~~~~~^^^^^^^^^^^^^
  File "/Users/gugu/Code/cpython/Lib/calendar.py", line 165, in monthrange
    raise IllegalMonthError(month)
calendar.IllegalMonthError: bad month number -1; must be 1-12

>>> calendar.month(2024, 13)
Traceback (most recent call last):
  File "<python-input-13>", line 1, in <module>
    calendar.month(2024, 13)
    ~~~~~~~~~~~~~~^^^^^^^^^^
  File "/Users/gugu/Code/cpython/Lib/calendar.py", line 390, in formatmonth
    s = self.formatmonthname(theyear, themonth, 7 * (w + 1) - 1)
  File "/Users/gugu/Code/cpython/Lib/calendar.py", line 373, in formatmonthname
    s = month_name[themonth]
        ~~~~~~~~~~^^^^^^^^^^
  File "/Users/gugu/Code/cpython/Lib/calendar.py", line 103, in __getitem__
    funcs = self._months[i]
            ~~~~~~~~~~~~^^^
IndexError: list index out of range
```

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

macOS

### Linked PRs
* gh-126484
* gh-126878
* gh-126879

### PR DESCRIPTION
**Pull request by Zheaoli**
Fix: #126476

* Issue: gh-126476

### CODE REVIEW

**Comment by ericvsmith** [comment]
I'm not sure this is a good idea. If someone was catching `IndexError`  previously, then the change to `IllegalMonthError` will break their code.

**Comment by ZeroIntensity** [comment]
Yeah, Eric is right. I would be ok with this if `IllegalMonthError` also inherited from `IndexError`.

**Comment by Zheaoli** [comment]
I have update this PR,PTAL @ericvsmith @ZeroIntensity

**Review by ZeroIntensity — COMMENTED** [review]
I'm not sure how to feel about this. It works, and the implementation looks fine, but it's also a little hacky so I'm hesitant to approve. I'll wait for Eric's input.

**Comment by ericvsmith** [comment]
Yeah, I'm not crazy about it. It seems like there must be other places where we've done something similar, though. Does anyone feel like searching the stdlib for some place else we've done this?

**Comment by tomasr8** [comment]
I didn't find any examples of errors inheriting from `IndexError`, but there is precedent for changing `IndexError` to something else. Here are two examples that I found:

https://github.com/python/cpython/pull/8184/files#diff-654faea1665f8db34a06e155e3f168b467e2071967404ea9692f41190fcbb5d4

https://github.com/python/cpython/pull/123613/files#diff-f6ee3a81bee278bb507d0cba22bb56acb204667c1271fc5d3fc41aa702248c35

**Comment by ericvsmith** [comment]
I was thinking more along the lines of using multiple inheritance for exceptions. But those are good examples of changing the type of an exception.

**Comment by tomasr8** [comment]
> I was thinking more along the lines of using multiple inheritance for exceptions.

Ah sorry, I misunderstood. I checked again and found two examples of multiple inheritance:

```python
class MultipartConversionError(MessageError, TypeError):
    """Conversion to a multipart is prohibited."""


class UnsupportedOperation(OSError, ValueError):
        pass
```

**Review by ethanfurman — APPROVED** [review]
Multiple inheritance for exceptions is a common technique (at least in my own code  :smile: ).  Looks good to me.

**Review by hauntsaninja — COMMENTED** [review]
Thanks Tomas for finding examples! I'd vote that it's probably okay to change the exception type. Adding `IndexError` as a base can break code too if someone has an `except IndexError` for some other reason and calls `monthrange`.

(side note: the `UnsupportedOperation` Tomas mentions has [finicky history](https://github.com/python/cpython/commit/606ab86c0e11eb00fc276c694bff9d7f9239d538))

File: Misc/NEWS.d/next/Library/2024-11-06-18-30-50.gh-issue-126476.F1wh3c.rst [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Raise :class:`calendar.IllegalMonthError` for :func:`calendar.month`
```
Comment (ethanfurman) [inline Misc/NEWS.d/next/Library/2024-11-06-18-30-50.gh-issue-126476.F1wh3c.rst]:
```suggestion
Raise :class:`calendar.IllegalMonthError` (now a subclass of :class:`IndexError`) for :func:`calendar.month`
```

**Pull request merged**